Attempt to use get storage model from ID_MODEL_ENC before trying ID_MODEL(#7665)

ltrager opened the pull request:
Signed-off-by: Lee Trager <lee.trager@canonical.com>

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

stgraber commented on lxd/resources/storage.go:
```diff
@@ -84,8 +84,17 @@ func storageAddDriveInfo(devicePath string, disk *api.ResourcesStorageDisk) erro
+		if udevProperties["E:ID_MODEL_ENC"] != "" {
+			// The encoded model may contain escaped hex values which need to
+			// be converted to ASCII
+			model, err := strconv.Unquote(`"` + udevProperties["E:ID_MODEL_ENC"] + `"`)
```
We only use ` for multi-line quoting and tend to prefer Sprintf over string concatenation.

stgraber replied:
Right, we can't use this logic to decode _ENC, there is a lot more to _ENC that this.

stgraber commented:
So it looks like we do need to use ID_MODEL_ENC and ID_VENDOR_ENC to obtain the original value from ATA/SCSI (when available). However to do this properly, we need the opposite of: https://github.com/systemd/systemd/blob/master/src/basic/device-nodes.c#L22

And hope that systemd won't change encoding methods.


To validate the decoder, we should check it against:
 - ASCII strings containing spaces, slashes, backslashes, #, +, ., lower and uppercase letters and digits
 - UTF-8 strings including multi-byte glyphs

Feeding the input to the udev C function and confirming the decoder returns byte for byte identical output.
The decoder itself shouldn't do any space stripping.

ltrager commented:
Thanks for the review. I've updated the patch to implement the inverse of encode_devnode_name. From my reading of the code while encode_devnode_name validates that unicode is properly encoded the only time encoding occurs is when a character is detected which isn't allowed in /dev. As such I've implemented decode_devnode_name to decode any detected encoded hex values. I'm not validating whether the decode value is on udev's deny list or not incase they change it in the future.

stgraber commented on lxd/resources/storage.go:
```diff
@@ -20,6 +21,28 @@ var devDiskByID = "/dev/disk/by-id"
 var runUdevData = "/run/udev/data"
 var sysClassBlock = "/sys/class/block"
 
```
Go prefers camelCase over snake_case, so decodeUdev or something along those lines would be better. It should also live in utils.go rather than in storage.go

ltrager commented:
Thanks for the review. Moved the new function to utils.decodeDevNodeName

stgraber commented:
Should be lxd/resources/utils.go

stgraber commented:
Fixed in place.

stgraber commented:
jenkins: test this please

stgraber commented:
Ah, had to de-snake-case variable names too to make CI happy.

stgraber commented:
jenkins: test this please

lxc-jenkins commented:
Testsuite passed
